kernel/posix/pthread_sched.c

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <posix_sched.h>
+
+int sched_get_priority_min(int policy)
+{
+	if (!((policy == SCHED_FIFO) || (policy == SCHED_RR))) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	if (IS_ENABLED(CONFIG_COOP_ENABLED)) {
+		if (policy == SCHED_FIFO) {
+			errno = 0;	/* Valid negative return */
```

Comment:
This function can't return a negative value on success; that's how the API is defined, and how code using it will expect it to behave.

A solution is not to have priorities mapping 1:1 between Zephyr and POSIX.  You can derive the formula for the proper mapping like so:

- Consider two points in the space: (x0, y0) and (x1, y1).  The X axis is the POSIX priority, and the Y axis is the Zephyr priority.
- Find two points, P1 and P2, that maps the minimum priority in Zephyr (that's a negative value) to a value that's acceptable in POSIX (say, 1). Do the same for P2, for the maximum value.
- Calculate the slope: m = (y1 - y0) / (x1 - x0)
- Use the calculated slope in the line equation: y - y0 = m(x - x0), where y0 and x0 are the coordinates from P1
- Isolate y or x depending on which direction you want to go (Zephyr -> POSIX or POSIX -> Zephyr).  If you're calculating y, any x in the line is valid (could be even x0), and vice versa.
- It's important to round m to the nearest integer, to not use floating point math here.  It shouldn't matter if it's ceiled or floored

Reply:
As per my understanding if a function returns a negative value then it should set errno to 0.
because of which I was putting 0 as ERROR code.

 https://www.unix.com/man-page/posix/3posix/sched_get_priority_min/

Reply:
That's the opposite: if the function returns -1, errno must be set; otherwise, errno isn't touched (see **RETURN VALUE** in the linked man page).

Per the POSIX specification, only a return value of -1 will signal error.  However, a lot of code will verify error conditions by testing the sign bit (i.e. checking if the return value is < 0 rather than == -1), as it often generates slightly tighter code.

Being a compatibility layer, it's best to assume that any function returning -1 per the specification will also return a negative number (whatever it is) on error, and 0 or a positive value on success.  Returning any value from 0 to INT_MAX is fine for these functions, as long as they're consistent with each other, and that the functions to set the priorities are also consistent with these priority values.

Reply:
Looking at documentation @ https://www.systutorials.com/docs/linux/man/2-sched_get_priority_max/#index, (Looks like man page for IEEE Std 1003.13-2003 is silent on number of priorities level  in POSIX https://www.unix.com/man-page/posix/3posix/sched_get_priority_min/)- 
As per POSIX :- 
1. Higher the numerical value Priority, Higher the priority. Something like a thread with priority 5 will have higher priority than a thread with priority 1. (something opposite of Zephyr).
 2. There should be at least 32 priority levels, but this can vary based on other standards. I think we can avoid this as for RTOS this may be too high number of priority levels.

I think we can not use the slope technique to map Zephyr priort levels to POSIX prioirty levels as this may create confusion/issues.  Let us assume (Numerical lower priority has Higher Priority than numberical high only for example). 
Let us take an example - If we have premptive based scheduling enabled with 
Zephyr_Max_ priority = 15
Zephyr_Min_ priority = 0
Now mapping this range to POSIX_MAX = 31, POSIX_MIN=0.
Let us say that a POSIX compliant application has 3 threads with priority 2, 3, 4. then it may be expected that   1 has higher priority than 2 and so on.
But internally :- 
Thread 1 ==> POSIX_PPRIOITY = 2 mapped to ZEPHYR_PRIORTY = 1
Thread 2 ==> POSIX_PPRIOITY = 3 mapped to ZEPHYR_PRIORTY = 1
Thread 3 ==> POSIX_PPRIOITY = 4 mapped to ZEPHYR_PRIORTY = 2
So this may give undesirable results.  So i think there should be 1:1 or n:m (n>= m) mapping between POSIX and ZEPHYR .

I think we can map ZEPHYR_PRIORITY to POSIX_PRIORTY :-
For Preemptive_thread :-  
posix_priorty= (CONFIG_NUM_PREEMPT_PRIORITIES - 1) - zephyr_priorty
*-1 to keep maximum priority reserved for highest priority thread.

This will make POSIX_MIN_PRIORITY = 0 and POSIX_MAX_PRIORTY = CONFIG_MAX_APPLICATION_PRIORITY
For COOPERATIVE scheduling  policy:- 
if preemption Enabled :- 
posix_application_priority = CONFIG_NUM_COOP_PRIORITIES  + zephyr_priority. (cooperative priorities are -ive).
if preemption Enabled :- 
posix_application_priority = (CONFIG_NUM_COOP_PRIORITIES - 1)  + zephyr_priority. (cooperative priorities are -ive).
*-1 to keep maximum priority reserved for highest priority thread.